1.0: Fluent, Deno support, docs, and more (#20)
grammyjs/i18n: README.md

# grammY i18n

Internationalization plugin for [grammY](https://grammy.dev) based on
[Project Fluent](https://projectfluent.org).

## Installation

#### Node.js

```sh
npm install @grammyjs/i18n
```

#### Deno

> 
> If you are using [Deno](https://deno.land), you can import the plugin from the
> Deno third-party module registry:

Reviewer: This is not necessary imho.

```suggestion
```

```ts
import { I18n, I18nContextFlavor } from "https://deno.land/x/.../mod.ts";
```

## Usage

> The following examples are written for [Deno](https://deno.land). You can still
> use the same code in Node.js by changing the imports accordingly. Checkout the
> [examples folder](examples/) for full examples of both Deno and Node.

Reviewer: ```suggestion
The following examples are written for [Deno](https://deno.land).
You can still use the same code in Node.js by changing the imports accordingly.
Check out the [examples/](examples/) for full examples of both Deno and Node.
```

To setup the translations quickly, first of all you need to put all of your
translation files in a folder (Or, see
[Adding Translations](#adding-translations)). Usually, we name the folder
**locales**. And the translation files' name should end with `.ftl` (fluent)

> extension. Here is an example project structure:

Reviewer: ```suggestion
To quickly setup the translations, first of all, you need to put all of your translation files in a directory (or see [Adding Translations](#adding-translations)).
Usually, the name of that folder is going to be **locales/**.
And the translation files name should have the extension `.ftl` (fluent).

Here is an example project structure:
```

```
.

> ├─ 📂 locales

Reviewer: (personal suggestion: the emoji will not look as expected everywhere. For example, on systems with no emoji fonts.)

```suggestion
├─ locales
```

│  ├── en.ftl
│  ├── it.ftl
│  └── ru.ftl
└── bot.ts
```

> > Have a look at the Fluent syntax guide if you need help with the `.ftl`
> > (fluent) files: https://projectfluent.org/fluent/guide.

Reviewer: ```suggestion
> Have a look at the Fluent syntax guide if you need help with the syntax of Fluent files:
> https://projectfluent.org/fluent/guide.
```

And in the `bot.ts` file:

> If you are using [TypeScript](https://typescriptlang.org) (recommended to use)
> make sure to use `I18nContextFlavor` to extend your bot context in order for

> > typings to work correctly.

Reviewer: ```suggestion
> If you are using [TypeScript](https://typescriptlang.org) (recommended), make sure to use `I18nContextFlavor` to extend your bot's context type so that you won't have type errors.
```

> # i18n

Reviewer 2: ```suggestion
# grammY i18n
```

Reviewer: @dcdunkan

Author: Done

Internationalization plugin for [grammY](https://grammy.dev) based on
[Project Fluent](https://projectfluent.org).

## Installation

#### Node.js

```sh
npm install @grammyjs/i18n
```

#### Deno

If you are using [Deno](https://deno.land), you can import the plugin from the
Deno third-party module registry:

```ts

> import { I18n, I18nContextFlavor } from "https://deno.land/x/.../mod.ts";

Reviewer 2: I would prefer to have `grammy_i18n` here, but I doubt that it will be possible due to limitations of /x. If it will not be possible to use that name, I suggestion going for `grammy_internationalization` or `grammy_l10n`.

Reviewer 3: Trying to resolve this

  await ctx.reply(ctx.t("greeting"));
});

// Start your bot
bot.start();
```

### Using sessions

If `useSession` is set to true, the plugin will try to read the
`__language_code` from user's session data and use it, before falling back to
language code received from the updates.

> **NOTE**: You have to register the `session` middleware before registering
> i18n's middleware. Otherwise, it won't be able to read or write from session.

```ts
import {
  Bot,
  Context,
  session,
  SessionFlavor,
} from "https://deno.land/x/grammy/mod.ts";
import { I18n, I18nContextFlavor } from "https://deno.land/x/.../mod.ts";

interface SessionData {
  __language_code: string;
}

type MyContext =
  & Context
  & SessionFlavor<SessionData>
  & I18nContextFlavor;

const i18n = new I18n({
  defaultLocale: "en",
  useSession: true, // whether get/set in session
});

// Load locales from the `locales` directory.
await i18n.loadLocalesDir("locales");

const bot = new Bot<MyContext>(""); // <- Put your bot token here

// Remember to register `session` middleware before registering
// middleware of the i18n instance.
bot.use(session({
  initial: () => {

>     return { __language_code: "en" };

Reviewer 2: Wouldn't it be better to not return any default lang code here?
```suggestion
    return { };
```
That way, the plugin can infer it automatically from `ctx.from.language_code` as a fallback to when no value is present on the session.

Reviewer 3: I also would add a fallback lang parameter, so there are 2 fallbacks

Author: @KnorpelSenf Done.

Author: @HeeroML What? Why? There is a `localeNegotiator` option in i18n config. Isn't that enough?

The locale is selected in the following order.

1. From custom `localeNegotiator`. (if any provided)
2. Reads from session (if it is enabled)
3. User's from.language_code (from TG updates)
4. Default locale.

### Load locales from a directory

This has to be the simplest way of adding translations. Just put them all in a
folder and load them like this:

```ts
await i18n.loadLocalesDir("locales");
```

### Load a single locale

From a file:

```ts
await i18n.loadLocale("en", {
  filePath: "path/to/en.ftl",
});
```

From source text:

```ts
await i18n.loadLocale("en", {
  source: `greeting = Hello { $first_name }!
language-set = Language has been set to English!`,
});
```

## Custom placeables and functions

This plugin injects some custom functions and placeables to your translations
and translation context.

### Placeables

#### first_name

Replacement for `ctx.from.first_name`. So, you don't need to pass it in the
context when you need it. You can just use it directly.

```fluent
greeting = Hi {$first_name}
```

### Functions

#### CTX

`CTX("key", "optionalDefaultValue")`

This function helps you to access any properties of the bot context directly
from your translations.

```fluent
id = Your user ID: {CTX("from.id")}
```

Optionally you can pass in a default value as the second argument. If the value
is undefined the default value is used instead.

> ## API Documentation

Reviewer 2: Why would this be here, rather than on `doc.deno.land`? The explanations could be used for the aforementioned JSDoc comments.

Author: Done.

4. Default locale from the configuration.
5. English (en).

Locale negotiation normally happens only once during Telegram update processing.
However, you can call `ctx.i18n.reNegotiateLocale()` to call the negotiator
again and determine the new locale. This is useful if the locale changes during
single update processing.

Here is an example locale negotiator:

```ts
async function myLocaleNegotiator(ctx: Context) {
  return ctx.session.locale ?? ctx.from?.language_code ?? "en";
}

const i18n = new I18n({
  localeNegotiator: myLocaleNegotiator,
});
```

## Credits

Thanks to...

- **Slava Fomin II** ([@slavafomin](https://github.com/slavafomin)) for the
  Node.js implementation of the
  [original Fluent plugin](https://github.com/the-moebius/grammy-fluent) and the
  [better Fluent integration](https://github.com/the-moebius/fluent).

- **Roj** ([@roj1512](https://github.com/roj1512)) for the
  [Deno port](https://github.com/roj1512/fluent) of the original
  [@fluent/bundle](https://github.com/projectfluent/fluent.js/tree/master/fluent-bundle)
  and
  [@fluent/langneg](https://github.com/projectfluent/fluent.js/tree/master/fluent-langneg)
  packages.

> - And all the previous maintainers and contributors of this i18n plugin.

Reviewer 2: @dcdunkan please add yourself in an appropriate position to this list, you contributed substantially and you belong here. This is especially true given that you do not intend to stay the primary maintainer, so you shall not go unmentioned in the time ahead :)
